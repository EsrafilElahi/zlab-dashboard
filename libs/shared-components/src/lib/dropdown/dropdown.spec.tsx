import {
  fireEvent,
  render,
} from '@testing-library/react';

import Dropdown from './dropdown';

const dummyData = [
  { id: 1, label: 'Kassulke & Sohn', value: 'Kassulke & Sohn' },
  { id: 2, label: 'Bayer-Bergnaum', value: 'Bayer-Bergnaum' },
  { id: 3, label: 'Schinner Group', value: 'Schinner Group' },
];

describe('Dropdown', () => {
  it('should render successfully', () => {
    const { baseElement } = render(
      <Dropdown
        onChange={(v) => {
          console.log(v);
        }}
        options={dummyData}
      />
    );
    expect(baseElement).toBeTruthy();
  });

  it('should match the snapshot', () => {
    const { baseElement } = render(
      <Dropdown
        onChange={(v) => {
          console.log(v);
        }}
        options={dummyData}
      />
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('Inner container should not have children at start', () => {
    const { getByTestId } = render(
      <Dropdown
        onChange={(v) => {
          console.log(v);
        }}
        options={dummyData}
      />
    );

    expect(getByTestId('dropdown-inner-container').hasChildNodes()).toBeFalsy();
  });

  it("Start Selected Option text should be equal to 'Select option'", () => {
    const { getByText } = render(
      <Dropdown
        onChange={(v) => {
          console.log(v);
        }}
        options={dummyData}
      />
    );
    const expectingText = 'Select option';
    expect(getByText(expectingText)).toBeTruthy();
  });

  it('Clicking on dropdown should expose its values', async () => {
    const { findByTestId, getByTestId } = render(
      <Dropdown
        onChange={(v) => {
          console.log(v);
        }}
        options={dummyData}
      />
    );

    fireEvent(
      getByTestId('parent-container'),
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      })
    );

    const innerContainer = await findByTestId('dropdown-inner-container');
    expect(
      innerContainer.querySelector('ul')?.childElementCount
    ).toBeGreaterThan(0);
  });

  it('Selected item should change after clicking on an item', async () => {
    const { findByTestId, getByTestId, getByText } = render(
      <Dropdown
        onChange={(v) => {
          console.log(v);
        }}
        options={dummyData}
      />
    );

    fireEvent(
      getByTestId('parent-container'),
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      })
    );

    const firstItemValue = dummyData[0].label;

    const innerContainer = await findByTestId('dropdown-inner-container');

    fireEvent(
      innerContainer.querySelectorAll('ul li')[0],
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      })
    );

    expect(getByText(firstItemValue)).toBeTruthy();
  });

  it('should match the snapshot when its open', () => {
    const { baseElement, getByTestId } = render(
      <Dropdown
        onChange={(v) => {
          console.log(v);
        }}
        options={dummyData}
      />
    );

    fireEvent(
      getByTestId('parent-container'),
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      })
    );

    expect(baseElement).toMatchSnapshot();
  });
});
