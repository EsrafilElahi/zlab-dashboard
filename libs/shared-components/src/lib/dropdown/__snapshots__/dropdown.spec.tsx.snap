// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown should match the snapshot 1`] = `
<body>
  <div>
    <div
      class="container"
    >
      <div
        class="selected-option-container"
        data-testid="parent-container"
      >
        <p
          class="selection-option"
        >
          Select option
        </p>
      </div>
      <div
        class="inner-container "
        data-testid="dropdown-inner-container"
      />
    </div>
  </div>
</body>
`;

exports[`Dropdown should match the snapshot when its open 1`] = `
<body>
  <div>
    <div
      class="container"
    >
      <div
        class="selected-option-container"
        data-testid="parent-container"
      >
        <p
          class="selection-option"
        >
          Select option
        </p>
      </div>
      <div
        class="inner-container active"
        data-testid="dropdown-inner-container"
      >
        <input
          class="search-input"
          placeholder="Search..."
          type="text"
          value=""
        />
        <ul
          class="options-list"
        >
          <li>
            Kassulke & Sohn
          </li>
          <li>
            Bayer-Bergnaum
          </li>
          <li>
            Schinner Group
          </li>
        </ul>
      </div>
    </div>
  </div>
</body>
`;
